File: scripts/generators/javascript/serve_site.js
#!/usr/bin/env node
/**
 * DocShell JavaScript (Node.js) Web Server & RAG API
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const url = require('url');

const rootDir = path.resolve(__dirname, '../../../');
const distWeb = path.join(rootDir, 'dist', 'webpage');
let port = 8000;

process.argv.forEach((arg, i) => {
  if ((arg === '-p' || arg === '--port') && process.argv[i + 1]) {
    port = parseInt(process.argv[i + 1], 10);
  }
});

const mimeTypes = {
  '.html': 'text/html; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.js': 'application/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.svg': 'image/svg+xml',
  '.ico': 'image/x-icon'
};

const server = http.createServer((req, res) => {
  const parsedUrl = url.parse(req.url, true);
  const pathname = parsedUrl.pathname;

  // CORS Headers
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    res.writeHead(200);
    res.end();
    return;
  }

  // API Status
  if (pathname === '/api/status' && req.method === 'GET') {
    res.writeHead(200, { 'Content-Type': 'application/json; charset=utf-8' });
    res.end(JSON.stringify({
      status: 'online',
      runtime: 'javascript (node.js)',
      node_version: process.version,
      rag_enabled: true
    }));
    return;
  }

  // API Chat / RAG
  if (pathname === '/api/chat' && req.method === 'POST') {
    let body = '';
    req.on('data', chunk => body += chunk);
    req.on('end', () => {
      try {
        const payload = JSON.parse(body);
        const query = (payload.message || '').toLowerCase();
        
        const indexPath = path.join(distWeb, 'search_index.json');
        let searchIndex = [];
        if (fs.existsSync(indexPath)) {
          searchIndex = JSON.parse(fs.readFileSync(indexPath, 'utf-8'));
        }

        const terms = query.split(/\s+/).filter(t => t.length > 2);
        let matches = [];

        searchIndex.forEach(item => {
          const text = (item.text + ' ' + item.chunk_title + ' ' + item.doc_title).toLowerCase();
          let score = 0;
          terms.forEach(t => {
            if (text.includes(t)) score++;
          });
          if (score > 0) matches.push({ score, item });
        });

        matches.sort((a, b) => b.score - a.score);

        if (matches.length > 0) {
          const top = matches[0].item;
          const resp = `**Resultado (Node.js RAG Engine):**\n\nNo documento **${top.doc_title}** (Seção: *${top.chunk_title}*):\n\n> ${top.text}`;
          res.writeHead(200, { 'Content-Type': 'application/json; charset=utf-8' });
          res.end(JSON.stringify({ response: resp, sources: [top.doc_title], runtime: 'node-rag' }));
        } else {
          res.writeHead(200, { 'Content-Type': 'application/json; charset=utf-8' });
          res.end(JSON.stringify({ response: 'Não encontrei trechos correspondentes nesta consulta.', sources: [], runtime: 'node-rag' }));
        }
      } catch (err) {
        res.writeHead(500, { 'Content-Type': 'application/json; charset=utf-8' });
        res.end(JSON.stringify({ error: err.message }));
      }
    });
    return;
  }

  // Servir arquivos estáticos
  let safePath = path.normalize(pathname).replace(/^(\.\.[\/\\])+/, '');
  let filePath = path.join(distWeb, safePath === '/' ? 'index.html' : safePath);

  fs.stat(filePath, (err, stats) => {
    if (err || !stats.isFile()) {
      res.writeHead(404, { 'Content-Type': 'text/plain; charset=utf-8' });
      res.end('404 Not Found - DocShell');
      return;
    }

    const ext = path.extname(filePath).toLowerCase();
    const contentType = mimeTypes[ext] || 'application/octet-stream';

    res.writeHead(200, { 'Content-Type': contentType });
    fs.createReadStream(filePath).pipe(res);
  });
});

console.log('=================================================================');
console.log('⚡ DocShell Node.js Server & API');
console.log(`   URL: http://127.0.0.1:${port}`);
console.log(`   Doc Root: ${distWeb}`);
console.log('   Pressione Ctrl+C para encerrar.');
console.log('=================================================================');

server.listen(port, '0.0.0.0');

File: scripts/generators/python/serve_site.py
#!/usr/bin/env python3
"""
DocShell Python Web Server & RAG API
Executa o servidor local HTTP para a documentação e fornece a API REST do Chatbot IA.
"""

from __future__ import annotations

import argparse
import json
import os
import sys
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from typing import Any
from urllib.parse import urlparse

def _resolve_root() -> Path:
    current = Path(__file__).resolve()
    for p in current.parents:
        if (p / "docs").exists() or (p / "publication").exists() or (p / "scripts").exists():
            return p
    return Path("/app") if Path("/app").exists() else current.parent

ROOT_DIR = _resolve_root()
if str(ROOT_DIR) not in sys.path:
    sys.path.insert(0, str(ROOT_DIR))

from scripts.core.config_loader import load_publication_config
from scripts.core.rag_engine import DocShellRAG


class DocShellHTTPHandler(SimpleHTTPRequestHandler):
    def __init__(self, *args, directory=None, **kwargs):
        self.rag_engine = DocShellRAG(ROOT_DIR)
        super().__init__(*args, directory=directory, **kwargs)

    def do_GET(self):
        parsed = urlparse(self.path)
        if parsed.path == "/api/status":
            self.send_response(200)
            self.send_header("Content-Type", "application/json; charset=utf-8")
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()
            status_data = {
                "status": "online",
                "runtime": "python",
                "rag_enabled": True,
                "ollama_host": self.rag_engine.ollama_host,
                "chat_model": getattr(self.rag_engine, "active_model", "llama3.2")
            }
            self.wfile.write(json.dumps(status_data).encode("utf-8"))
            return

        super().do_GET()

    def do_POST(self):
        parsed = urlparse(self.path)
        if parsed.path == "/api/chat":
            content_length = int(self.headers.get("Content-Length", 0))
            body_bytes = self.rfile.read(content_length)
            
            try:
                body_json = json.loads(body_bytes.decode("utf-8"))
                user_msg = body_json.get("message", "").strip() or body_json.get("query", "").strip()
                user_locale = body_json.get("locale", "pt-BR")
                
                result = self.rag_engine.ask(user_msg, locale=user_locale)
                
                self.send_response(200)
                self.send_header("Content-Type", "application/json; charset=utf-8")
                self.send_header("Access-Control-Allow-Origin", "*")
                self.end_headers()
                self.wfile.write(json.dumps(result, ensure_ascii=False).encode("utf-8"))
                return

            except Exception as e:
                self.send_response(500)
                self.send_header("Content-Type", "application/json; charset=utf-8")
                self.send_header("Access-Control-Allow-Origin", "*")
                self.end_headers()
                self.wfile.write(json.dumps({"error": str(e)}).encode("utf-8"))
                return

        if parsed.path == "/api/translate":
            content_length = int(self.headers.get("Content-Length", 0))
            body_bytes = self.rfile.read(content_length)
            try:
                body_json = json.loads(body_bytes.decode("utf-8"))
                text = body_json.get("text", "")
                target_loc = body_json.get("target_locale", "en-US")
                src_loc = body_json.get("source_locale", "pt-BR")

                from scripts.core.translator import translate_text
                translated = translate_text(text, target_locale=target_loc, source_locale=src_loc, root_dir=ROOT_DIR)

                self.send_response(200)
                self.send_header("Content-Type", "application/json; charset=utf-8")
                self.send_header("Access-Control-Allow-Origin", "*")
                self.end_headers()
                self.wfile.write(json.dumps({"translated_text": translated}, ensure_ascii=False).encode("utf-8"))
                return
            except Exception as e:
                self.send_response(500)
                self.send_header("Content-Type", "application/json; charset=utf-8")
                self.send_header("Access-Control-Allow-Origin", "*")
                self.end_headers()
                self.wfile.write(json.dumps({"error": str(e)}).encode("utf-8"))
                return

        self.send_error(404, "Endpoint não encontrado")

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.end_headers()


def run_server(port: int = 8000):
    dist_web = ROOT_DIR / "dist" / "webpage"
    if not (dist_web / "index.html").exists():
        print("⚠️ dist/webpage/index.html não encontrado. Compilando site...")
        from scripts.generators.python.build_site import build_python_site
        build_python_site()

    print("=================================================================")
    print(f"🚀 DocShell Python Server & RAG Engine")
    print(f"   URL: http://127.0.0.1:{port}")
    print(f"   Doc Root: {dist_web}")
    print("   Pressione Ctrl+C para encerrar.")
    print("=================================================================")

    handler = lambda *args, **kwargs: DocShellHTTPHandler(*args, directory=str(dist_web), **kwargs)
    server = ThreadingHTTPServer(("0.0.0.0", port), handler)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\n🛑 Servidor encerrado.")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="DocShell Python Web Server")
    parser.add_argument("-p", "--port", type=int, default=8000, help="Porta HTTP")
    args = parser.parse_args()
    run_server(args.port)
